lib/SILOptimizer/Utils/ConstExpr.cpp

```diff
@@ -0,0 +1,730 @@
+    auto operand2 = getConstantValue(inst->getOperand(2));
+    if (!operand0.isConstant())
+      return operand0;
+    if (!operand1.isConstant())
+      return operand1;
+    if (!operand2.isConstant())
+      return operand2;
+
+    // Overflowing integer operations like sadd_with_overflow take three
+    // operands: the last one is a "should report overflow" bit.
+    auto constFoldIntOverflow =
+        [&](const std::function<APInt(const APInt &, const APInt &, bool &)>
+                &fn) -> SymbolicValue {
+      if (operand0.getKind() != SymbolicValue::Integer ||
+          operand1.getKind() != SymbolicValue::Integer ||
+          operand2.getKind() != SymbolicValue::Integer)
+        return unknownResult();
+
+      auto l = operand0.getIntegerValue(), r = operand1.getIntegerValue();
+      bool overflowed = false;
+      auto result = fn(l, r, overflowed);
+
+      // Return a statically diagnosed overflow if the operation is supposed to
+      // trap on overflow.
+      if (overflowed && !operand2.getIntegerValue().isNullValue())
+        return evaluator.getUnknown(SILValue(inst), UnknownReason::Overflow);
+
+      auto &allocator = evaluator.getAllocator();
+      // Build the Symbolic value result for our normal and overflow bit.
+      return SymbolicValue::getAggregate(
+          {SymbolicValue::getInteger(result, allocator),
+           SymbolicValue::getInteger(APInt(1, overflowed), allocator)},
+          allocator);
+    };
+
+    switch (builtin.ID) {
+    default:
+      break;
+
+#define INT_OVERFLOW(OPCODE, METHOD)                                           \
+  case BuiltinValueKind::OPCODE:                                               \
+    return constFoldIntOverflow(                                               \
+        [&](const APInt &l, const APInt &r, bool &overflowed) -> APInt {       \
+          return l.METHOD(r, overflowed);                                      \
+        })
+      INT_OVERFLOW(SAddOver, sadd_ov);
+      INT_OVERFLOW(UAddOver, uadd_ov);
+      INT_OVERFLOW(SSubOver, ssub_ov);
+      INT_OVERFLOW(USubOver, usub_ov);
+      INT_OVERFLOW(SMulOver, smul_ov);
+      INT_OVERFLOW(UMulOver, umul_ov);
+#undef INT_OVERFLOW
+    }
+  }
+
+  LLVM_DEBUG(llvm::dbgs() << "ConstExpr Unknown Builtin: " << *inst << "\n");
+
+  // Otherwise, we don't know how to handle this builtin.
+  return unknownResult();
+}
+
+// Handle calls to opaque callees, either by handling them and returning None or
+// by returning with a Unknown indicating a failure.
+llvm::Optional<SymbolicValue>
+ConstExprFunctionState::computeOpaqueCallResult(ApplyInst *apply,
+                                                SILFunction *callee) {
+  LLVM_DEBUG(llvm::dbgs() << "ConstExpr Opaque Callee: " << *callee << "\n");
+  return evaluator.getUnknown((SILInstruction *)apply, UnknownReason::Default);
+}
+
+// TODO: Refactor this to someplace common, this is defined in Devirtualize.cpp.
```

Comment:
@marcrasi  This comment is a left-over from other patches right? It should not be removed.

Reply:
yeah, it shouldn't be in this patch. will remove.